cmd/cmd.go

```diff
@@ -70,3 +80,327 @@ func validateUserConfig(cfg *viper.Viper) error {
+type downloadWriter struct {
+	*download
+}
+
+// writeMetadata writes the exercise metadata.
+func (d downloadWriter) writeMetadata() error {
+	metadata := d.metadata()
+	return metadata.Write(d.destination())
+}
+
+// writeSolutionFiles attempts to write each exercise file that is part of the downloaded Solution.
+// An HTTP request is made using each filename and failed responses are swallowed.
+// All successful file responses are written except when 0 Content-Length.
+func (d downloadWriter) writeSolutionFiles() error {
+	if d.fromExercise {
+		return errors.New("existing exercise files should not be overwritten")
+	}
+	for _, filename := range d.Solution.Files {
+		res, err := d.requestFile(filename)
+		if err != nil {
+			return err
+		}
+		if res == nil {
+			continue
+		}
+		defer res.Body.Close()
+
+		// TODO: if there's a collision, interactively resolve (show diff, ask if overwrite).
+		// TODO: handle --force flag to overwrite without asking.
+
+		sanitizedPath := sanitizeLegacyNumericSuffixFilepath(filename, d.exercise().Slug)
+		fileWritePath := filepath.Join(d.destination(), sanitizedPath)
+		if err = os.MkdirAll(filepath.Dir(fileWritePath), os.FileMode(0755)); err != nil {
+			return err
+		}
+
+		f, err := os.Create(fileWritePath)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if _, err := io.Copy(f, res.Body); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// destination is the download destination path.
+func (d downloadWriter) destination() string {
+	return d.exercise().MetadataDir()
+}
+
+// downloadParams is required to create a download.
+type downloadParams struct {
+	usrCfg *viper.Viper
+	uuid   string
+	slug   string
+	track  string
+	team   string
+
+	fromExercise bool
+	fromFlags    bool
+}
+
+func newDownloadParamsFromExercise(usrCfg *viper.Viper, exercise workspace.Exercise) (*downloadParams, error) {
+	d := &downloadParams{
+		usrCfg:       usrCfg,
+		slug:         exercise.Slug,
+		track:        exercise.Track,
+		fromExercise: true,
+	}
+	return d, d.validate()
+}
+
+func newDownloadParamsFromFlags(usrCfg *viper.Viper, flags *pflag.FlagSet) (*downloadParams, error) {
+	d := &downloadParams{usrCfg: usrCfg, fromFlags: true}
+	var err error
+	d.uuid, err = flags.GetString("uuid")
+	if err != nil {
+		return nil, err
+	}
+	d.slug, err = flags.GetString("exercise")
+	if err != nil {
+		return nil, err
+	}
+	d.track, err = flags.GetString("track")
+	if err != nil {
+		return nil, err
+	}
+	d.team, err = flags.GetString("team")
+	if err != nil {
+		return nil, err
+	}
+	return d, d.validate()
+}
+
+// validate validates the presence of required downloadParams.
+func (d *downloadParams) validate() error {
+	if d.slug != "" && d.uuid != "" || d.uuid == d.slug {
+		if d.fromFlags {
+			return errors.New("need an --exercise name or a solution --uuid")
+		}
+		return errors.New("need a 'slug' or a 'uuid'")
+	}
+	if err := validateUserConfig(d.usrCfg); err != nil {
```

Comment:
`submit` and `download` require the same config values. I'm wondering if it'd be better to use config.Config (instead of viper.Viper) and move the validation directly to the config object. This strikes me as a case of Tell, Don't Ask.